woocommerce/woocommerce
# Fix caching of order refunds data

### Changes proposed in this Pull Request:

This pull request:

1. Introduces a new `OrderDataCache` class intended to cache order related data that isn't directly available in the `WC_Order` object.
2. Modifies the `WC_Order` object to use `OrderDataCache`, instead of directly using the WordPress cache, to store refunds related information.
3. Modifies `OrderCache` so that every time an order is inserted, updated or removed in/from the cache, the corresponding entry in `OrderDataCache` is deleted.
4. Modifies `OrdersTableRefundDataStore` and `WC_Order_Refund_Data_Store_CPT` so that every time a refund is created, updated or deleted, the corresponding entry in `OrderDataCache` (for the order that holds the refund) is deleted.
5. Introduces a new `InMemoryCacheEngine` class, intended for use in unit tests only.

Closes #38688.


### How to test the changes in this Pull Request:

To test the changes in this pull request you need a WordPress instance with an external object cache active (you can test if this is the case in WooCommerce - Status - "WordPress environment" section, "External object cache" key). An easy way to add such a cache is installing [the Docket Cache](https://wordpress.org/plugins/docket-cache/) plugin.

1. Set the new orders table as authoritative in WooCommerce - Settings - Advanced - Features - "Data storage for orders". Turn also sync on to ease testing, although in principle it shouldn't be necessary.
2. Go to Orders and create a new order.
3. Add a fee of any value to the order ("Refund" button) and save the order ("Update" button).
4. Add a refund for any amount that is less than the value of the refund ("Refund" button, enter the amount, click "Refund X manually" button), verify that the refund is created correctly and that the "Refunded" and "Net payment" amounts are correct.
5. Delete the refund clicking the X icon next to it, verify that the refund isn't listed anymore and that the "Refunded" and "Net payment" amounts are still correct.
7. Repeat, this time setting the posts table as authoritative in step 1.

### Changelog entry




<details>

#### Significance




#### Type




#### Message

Fix caching of refund data by introducing a new order data cache

#### Comment

</details>
